Carrier management system

gabrielleungkh opened the pull request:
Here is the CarrierManagementSystem code which mainly uses the builder pattern to create carriers. Please review and give comments!

RPahlevan committed:
Carrier Management System

Its a first prototype.
The code needs some clean up.

john-breton committed:
Add Maven to carriermanagementsystem

john-breton committed:
Merge pull request #1 from RPahlevan/temp-maven

Temp maven

john-breton committed:
Remove erroneous package

RPahlevan committed:
Carrier Management System code clean up

john-breton committed:
Refactor Code

- Tried to incorporate lambda expressions where possible (still a few places they could be added but oh well).
- Remove some unnecessary checks in the RFPorts and CarrierFrequencies enums.
- Corrected typos and removed redundant spaces
- Used StringBuffers instead of Strings to keep multi-threading support possible if it comes up
- Used System.arrayCopy() whenever manual for loops were being used to copy one array to another.

john-breton committed:
Merge branch 'CarrierManagementSystem' of https://github.com/RPahlevan/OODesignFinal2021 into CarrierManagementSystem

john-breton committed:
Merge Changes

- Reverted the names of certain fields from all uppercase to camelCase (uppercase fields need to be final AND static per the Java style guidelines. Here they were only final).
- Cleaned up some of the Javadoc.
- Converted additional uses of arrays to Lists to maintain consistency.
- Threw in some streams to make Ista happier.

gabrielleungkh committed:
Added Carrier class, randomized unique carrierId assignment, name refactoring

- Added Carrier abstract class that is now extended by LteCarrier and WcdmaCarrier and most things from these two classes have been abstracted out
- Added CarrierManagementIf which is the front-facing API for the carrier management system. It is implemented by the CarrierManagementSystemDirector class which manages the carrier builder classes.
- Added a constructor to CarrierManagementSystemDirector which will now instantiate a stack of random integers from 1-1000 to be used for assigning unique carrierId's to each carrier to be created. createLteCarrier and createWcdmaCarrier no longer take in carrierId as input as a result
- createLteCarrier and createWcdmaCarrier has return type of Carrier, but each returns the specific subclass type respectively
- Refactored the naming convention of classes, methods, and some variables to match RU side (i.e. LTECarrier -> LteCarrier)

RPahlevan committed:
some minor change into Carrier management to better use the abstract class and id generator stack.

gabrielleungkh committed:
Make CarrierManagementSystemDirector a singleton

- Made the CarrierManagementSystemDirector class a singleton with lazy instantiation and thread safety
- Updated CarrierManagementSystemClient to use the singleton
- Small formatting changes

dLeung09 commented on src/main/java/carrierManagementSystem/Carrier.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return The collected information, as a String.
+     */
+    public abstract String print();
```
Minor - for convention's sake, we could just override the `toString()` method

dLeung09 commented on src/main/java/carrierManagementSystem/Carrier.java:
```diff
@@ -0,0 +1,68 @@
+ * This abstract class defines a Carrier that will have concrete implementations
+ * for different RAT types.
+ */
+public abstract class Carrier {
```
By the looks of it, this is an immutable class - implicitly this means if we were to want to modify a carrier (e.g. change it's BandFrequency), it would involve creating a new carrier and deleting.
As far as I can tell, BandFrequency is the only attribute of a carrier that can change after it is created, so perhaps we should add a setter here? The other ones are optional (in my mind), given there doesn't appear to be a requirement that would force us to change (say, for example) the RF Port, but might be something we'd need in the future.

dLeung09 replied:
I realize I used the term `BandFrequency`, where you've called it `CarrierFrequencies` - they're the same thing, just a miscommunication on my part.

RPahlevan replied:
The carrier class sits on top of lteCarrier and WcdmaCarrier class. Normally the concrete classes just create an object  and return it to relevant builder class. The concrete class itself does not have any setter method and all the setter method should implemented inside the builder class. That's why in our CarrierBuilder class, we have several setter methods.

dLeung09 commented on src/main/java/carrierManagementSystem/Carrier.java:
```diff
@@ -0,0 +1,68 @@
+    protected CarrierFrequencies carrierFrequencies;
+    protected Double transmittingPower;
+
+    public Carrier(Integer carrierId, List<RfPorts> rfPorts, CarrierFrequencies carrierFrequencies,
```
Do we want more validation here? What happens if we're given an invalid `CarrierFrequencies` value? If `rfPorts` is `null`, what happens? What about if it's empty?

RPahlevan replied:
We already address this type of issue in our builder method. An appropriate exception will be threw based on the input.

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierBuilder.java:
```diff
@@ -0,0 +1,42 @@
+public interface CarrierBuilder {
+
+    /**
+     * Sets up an ID for both LTE and WCDMA carrier.
```
Very minor: we likely don't want to mention `LTE` and `WCDMA` explicitly here - if a future requirement introduces a new RAT type this comment could lead to some confusion. Referring to it as a carrier (or generic carrier) would probably be sufficient

RPahlevan replied:
I am agree with you and just changed it.

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierFrequencies.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * This class contain information related to LTE and WCDMA bands.
+ */
+public enum CarrierFrequencies {
```
Minor: Could we make this singular? Plural seems to indicate that each of these enumerations carries multiple values
`CarrierFrequency`

dLeung09 replied:
I also wouldn't be opposed to changing the name to `FrequencyBand`, since that's what we've named the underlying string value for the enumeration :)

RPahlevan replied:
good suggestion, done !

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierFrequencies.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final String FrequencyBand;
+
+    CarrierFrequencies(String FrequencyBand) {
```
Should probably be `private`

RPahlevan replied:
agreed! done

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemClient.java:
```diff
@@ -0,0 +1,41 @@
+ * Client class. This class would perform the unit tests for the carrier
+ * management system.
+ */
+public class CarrierManagementSystemClient {
```
Should this cover some error cases? For example: creating an WCDMA carrier with 4 ports, creating an LTE carrier using a WCDMA frequency, etc.

RPahlevan replied:
yes, good point. 
I add some cases into the client class for the demo purpose.

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class CarrierManagementSystemDirector implements CarrierManagementIf {
+    private static volatile CarrierManagementSystemDirector UNIQUE_INSTANCE = null;
+    private Stack<Integer> carrierIdGenerator = new Stack<Integer>();
+    private ArrayList<Integer> randomizeCarrierId = new ArrayList<Integer>();
```
`randomizeCarrierId` doesn't seem to be used outside of the constructor - is it necessary to persist it as a member variable for the class?

RPahlevan replied:
yes, I just fixed it.

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+    private ArrayList<Integer> randomizeCarrierId = new ArrayList<Integer>();
+
+    private CarrierManagementSystemDirector() {
+        for (int i = 1; i < 1001; i++) {
```
Minor: Can we make `1001` some constant somewhere, instead of a magic value here? I'd also suggest making it `1000` (and adjust the conditional appropriately) since the logic implies there are a max of `1000` carriers.

RPahlevan replied:
done

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            lteCarrier.setCarrierId(carrierIdGenerator.pop());
+        } catch (EmptyStackException e) {
+            System.out.println("carrierIdGenerator stack for LTE carrier is empty!");
```
Maybe a more user-friendly message is appropriate here?
`"Maximum LTE carriers reached. Please remove unused carriers"` or something similar?

RPahlevan replied:
thanks, done!

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+        lteCarrier.setFrequencyBand(carrierFrequencies);
+        lteCarrier.setTransmittingPower(transmittingPower);
+        lteCarrier.setRfPorts(rfPorts);
+        Carrier LTE = lteCarrier.getLteCarrier();
```
Uppercase variable names are conventionally reserved for constants - https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

Might consider changing `UNIQUE_INSTANCE` as well, however I can kinda see an argument for that one

john-breton replied:
UNIQUE_INSTANCE is iffy. Technically it should be static final, but Ista insisted during the midterm that we used volatile (even though final is technically a strong modifier but whatever). 
Regardless, I feel in this case it's fine to leave it capitalized.

RPahlevan replied:
done!

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+        try {
+            wcdmaCarrier.setCarrierId(carrierIdGenerator.pop());
+        } catch (EmptyStackException e) {
+            System.out.println("carrierIdGenerator stack for WCDMA carrier is empty!");
```
This message (and the equivalent LTE one) seem to imply that there are two separate ID generators for LTE and WCDMA, however that doesn't seem to be the case.

john-breton replied:
Indeed it seems like there is a hard total of 1000 for both WCDMA and LTE. Maybe we could use a separate stack for each so that we don't disproportionately have more WCDMA carriers compared to LTE carriers and vice versa?

RPahlevan replied:
done!

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+        wcdmaCarrier.setFrequencyBand(carrierFrequencies);
+        wcdmaCarrier.setTransmittingPower(transmittingPower);
+        wcdmaCarrier.setRfPorts(rfPorts);
+        Carrier WCDMA = wcdmaCarrier.getWcdmaCarrier();
```
See comment regarding `LTE` variable name above

RPahlevan replied:
done!

dLeung09 commented on src/main/java/carrierManagementSystem/RfPorts.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * This class contain information related to rf ports.
+ */
+public enum RfPorts {
```
Same comment about singular vs plural

RPahlevan replied:
I guess it should be plural as we configure several ports for both RAT type.

dLeung09 commented on src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java:
```diff
@@ -0,0 +1,74 @@
+ * will also assign a random unique carrier ID (in range 1 - 1000) to each
+ * carrier it creates.
+ */
+public class CarrierManagementSystemDirector implements CarrierManagementIf {
```
The two methods in this class `createLteCarrier()` and `createWcdmaCarrier()` are almost identical - the two main differences between them are the use of which builder instantiated and used (`LteCarrierBuilder` and `WcdmaCarrierBuilder` classes) and the method we call on that class (`getLteCarrier()` and `getWcdmaCarrier()`, which could be named more generically and put in the `CarrierBuilder` interface). Since both classes implement the `CarrierBuilder` interface (which, as a side note, isn't referenced anywhere), could some refactoring be done?
In my mind, this class would have a member variable of type `CarrierBuilder` and each of the public methods (`createLteCarrier` and `createWcdmaCarrier`) would set that member variable to the appropriate concrete class, then call a common method that calls each of the steps for creating the `Carrier`. This would also involve adding a public method to the `CarrierBuilder` class called `getCarrier()`, that could also be called in that common method. Let me know if you'd like me to explain this further (since this comment is getting rather long, and might not aptly describing what it is intended to)

RPahlevan replied:
we may need to talk more about it  however I made some changes into interface to make it more generic.

dLeung09 reviewed (commented):
Looks good! Added some comments

john-breton commented:
I think if most of David's comment can be addressed before Sunday we can move to get this merged such that it will require little to no changes for the next iteration. That being said if things can't be addressed, I move to merge this so we can at least have something ready for Monday.

RPahlevan committed:
implement the issue that was addressed

I made some changes into the code to fix all issue you were addressed in the code review.

RPahlevan commented:
Hello all,

Thanks for the code review and sorry for the late reply! 
I just update the code and submit my changes.
